src/dataflow/src/render/threshold.rs

```diff
@@ -20,56 +20,154 @@ use repr::{Diff, Row};
 use crate::render::context::CollectionBundle;
 use crate::render::context::{ArrangementFlavor, Context};
 
+/// A plan describing how to compute a threshold operation.
+#[derive(Debug)]
+pub enum ThresholdPlan {
+    /// Basic threshold maintains all positive inputs.
+    Basic(BasicThresholdPlan),
+    /// Retractions threshold maintains all negative inputs.
+    Retractions(RetractionsThresholdPlan),
+}
+
+/// A plan to maintain all inputs with positive counts.
+#[derive(Debug)]
+pub struct BasicThresholdPlan {
+    /// The number of columns in the input and output.
+    arity: usize,
+}
+
+/// A plan to maintain all inputs with negative counts, which are subtracted from the output
+/// in order to maintain an equivalent collection compared to [BasicThresholdPlan].
+#[derive(Debug)]
+pub struct RetractionsThresholdPlan {
+    /// The number of columns in the input and output.
+    arity: usize,
+}
+
+impl ThresholdPlan {
+    /// Construct the plan from the number of columns (`arity`). `maintain_retractions` allows to
+    /// switch between an implementation that maintains rows with negative counts (`true`), or
+    /// rows with positive counts (`false`).
+    pub fn create_from(arity: usize, maintain_retractions: bool) -> Self {
+        if maintain_retractions {
+            ThresholdPlan::Retractions(RetractionsThresholdPlan { arity })
+        } else {
+            ThresholdPlan::Basic(BasicThresholdPlan { arity })
+        }
+    }
+}
+
+/// Shared function to compute an arrangement of values matching `logic`.
+fn threshold_arrangement<G, T, R, L>(
+    arrangement: &R,
+    name: &str,
+    logic: L,
+) -> Arranged<G, TraceAgent<OrdValSpine<Row, Row, G::Timestamp, Diff>>>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+    R: ReduceCore<G, Row, Row, Diff>,
+    L: Fn(&isize) -> bool + 'static,
+{
+    arrangement.reduce_abelian(name, move |_k, s, t| {
+        for (record, count) in s.iter() {
+            if logic(count) {
+                t.push(((*record).clone(), *count));
+            }
+        }
+    })
+}
+
+pub fn build_threshold_basic<G, T>(
+    input: CollectionBundle<G, Row, T>,
+    arity: usize,
+) -> CollectionBundle<G, Row, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    // Arrange the input by all columns in order.
+    // Different trace variants require different implementations because their
+    // types are different, but the logic is identical.
```

Comment:
Is this vestigial from when there was a local function?

Reply:
Removed.